jdk-annotations/astubx-generator-cli/src/main/java/com/uber/nullaway/jdkannotations/AstubxGeneratorCLI.java

```diff
@@ -0,0 +1,251 @@
+    return modelData;
+  }
+
+  private static Map<String, List<ClassJson>> parseJson(String jsonDirPath) {
+    // get parsed JSON file
+    File jsonDir = new File(jsonDirPath);
+
+    if (!jsonDir.exists() || !jsonDir.isDirectory()) {
+      throw new IllegalArgumentException(
+          "JSON directory does not exist or is not a directory: " + jsonDirPath);
+    }
+
+    File[] jsonFiles = jsonDir.listFiles((dir, name) -> name.endsWith(".json"));
+    if (jsonFiles == null || jsonFiles.length == 0) {
+      throw new IllegalStateException("No JSON files found in: " + jsonDirPath);
+    }
+
+    Gson gson = new Gson();
+    Type parsedType = new TypeToken<Map<String, List<ClassJson>>>() {}.getType();
+
+    File jsonFile = jsonFiles[0]; // only one JSON file created
+
+    // parse JSON file
+    Map<String, List<ClassJson>> parsed;
+    try {
+      String jsonContent = Files.readString(jsonFile.toPath());
+      parsed = gson.fromJson(jsonContent, parsedType);
+    } catch (IOException e) {
+      System.err.println("Error reading JSON file: " + jsonFile.getAbsolutePath());
+      throw new RuntimeException(e);
+    }
+
+    return parsed;
+  }
+
+  private static void getMethodRecords(
+      ClassJson clazz,
+      String fullyQualifiedClassName,
+      Map<String, MethodAnnotationsRecord> methodRecords) {
+    for (MethodJson method : clazz.methods()) {
+      String methodName = method.name();
+      // skip all constructors
+      if (methodName.substring(0, methodName.indexOf('(')).equals(clazz.name())) {
+        continue;
+      }
+      // get return type nullness
+      String returnType = method.returnType();
+      ImmutableSet<String> returnTypeNullness = ImmutableSet.of();
+      // check if return type has Nullable annotation
+      if (returnType.contains("@org.jspecify.annotations.Nullable")) {
+        returnType = returnType.replace("@org.jspecify.annotations.Nullable ", "");
+        returnType = returnType.replace(" ", ""); // remove whitespace in Array types
+        returnTypeNullness = ImmutableSet.of("Nullable");
+      } else {
+        // check upperbound if return type is generic
+        for (int idx = 0; idx < clazz.typeParams().size(); idx++) {
+          if (returnType.equals(clazz.typeParams().get(idx).name())) {
+            returnTypeNullness = ImmutableSet.of("Nullable");
+          }
+        }
+      }
+      String signatureForMethodRecords = fullyQualifiedClassName + ":" + returnType + " ";
+      signatureForMethodRecords += methodName.substring(0, methodName.indexOf('(') + 1);
+      Map<Integer, ImmutableSet<String>> argAnnotation = new LinkedHashMap<>();
+
+      // get the arguments
+      String argsOnly = methodName.replaceAll(".*\\((.*)\\)", "$1").trim();
+      // split using comma but not the commas separating the generics
+      String[] arguments =
+          argsOnly.isEmpty() ? new String[0] : argsOnly.split(",(?=(?:[^<]*<[^>]*>)*[^>]*$)");
```

Comment:
That's one fancy regular expression!  I think at the least we need some comments to explain it.